### DISCUSSION TITLE
为什么你应该使用 Git 进行版本控制 · Why's THE Design? · /whys-the-design-git-the-best

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#573] [Announcements]
https://draveness.me/whys-the-design-git-the-best

**Comment by Maecenas** [comment]
这个月Hacker News上也有[关于Git的讨论](https://news.ycombinator.com/item?id=21418781)：Git的美妙之处来自于它的数据结构。

> As a result, you get lock-free transactions (branches) and minimal overhead. And through git‘s pointer-to-parent commit you get full lineage.

**Comment by draveness** [comment]
> @Maecenas
> 这个月Hacker News上也有[关于Git的讨论](https://news.ycombinator.com/item?id=21418781)：Git的美妙之处来自于它的数据结构。
> 
> > As a result, you get lock-free transactions (branches) and minimal overhead. And through git‘s pointer-to-parent commit you get full lineage.

Git 的数据结构确实非常有意思，有机会可以讲一讲

**Comment by xaclincoln** [comment]
在工作中，没有直接感觉到Git的可靠性（SVN也挺可靠的），在分支操作上能够明显察觉Git的高性能。有一点不好的是，Git的工作流相对复杂，需要培训和一段时间的适应，很多新人难以掌握。

**Comment by float2net** [comment]
对于这段描述不是很理解：
“在一些代码需要严格控制的项目中，集中式的版本控制系统却能更好的工作，这是因为企业或者项目在做决策时认为『代码需要非常严格的控制』”
为什么SVN相比于GIT能够更好的严格控制项目呢？

**Comment by draveness** [comment]
> 在工作中，没有直接感觉到Git的可靠性（SVN也挺可靠的），在分支操作上能够明显察觉Git的高性能。有一点不好的是，Git的工作流相对复杂，需要培训和一段时间的适应，很多新人难以掌握。

我认为这种培训和时间上的投入是值得的

**Comment by draveness** [comment]
> 对于这段描述不是很理解：
> “在一些代码需要严格控制的项目中，集中式的版本控制系统却能更好的工作，这是因为企业或者项目在做决策时认为『代码需要非常严格的控制』”
> 为什么SVN相比于GIT能够更好的严格控制项目呢？

SVN 是可以对目录划分权限的，在 Git 里面想要做到这点就稍微有些麻烦，我们需要分多个项目，然后用 submodule 来做这个事情。

**Comment by wujunze** [comment]
Git 是开发者的瑞士军刀

**Comment by Moligaloo** [comment]
这篇文章主要是拿集中式的 svn 与 git 做比较，那 git 相比其他的分布式版本管理软件（如 Mercurial）又有何有点和不足呢？

**Comment by draveness** [comment]
> @Moligaloo
> 这篇文章主要是拿集中式的 svn 与 git 做比较，那 git 相比其他的分布式版本管理软件（如 Mercurial）又有何有点和不足呢？

Mercurial 和 Git 都是分布式的版本管理系统，这两者从根源上的设计就决定了它们的差异不会特别大，区别也应该是在一些小功能或者实现细节上。

**Comment by Dream4ever** [comment]
感谢作者的文章，又长知识了。

PS：“Git 对每一个版本的提交都会生成一个 20 字节的 SHA-1 哈希”，这里应该是“40 字节”吧？

**Comment by draveness** [comment]
> 感谢作者的文章，又长知识了。
> 
> PS：“Git 对每一个版本的提交都会生成一个 20 字节的 SHA-1 哈希”，这里应该是“40 字节”吧？

是 20 字节啊 160bits == 20bytes https://en.wikipedia.org/wiki/SHA-1

**Reply by unbreakablehf** [reply]
git-doc/gitrevisions.html

The full SHA-1 object name (40-byte hexadecimal string), or a leading substring that is unique within the repository. E.g. dae86e1950b1277e545cee180551750029cfe735 and dae86e both name the same commit object if there is no other object in your repository whose object name starts with dae86e.

**Reply by draveness** [reply]
额，这里你想表达的意思是？

**Reply by unbreakablehf** [reply]
需要40个字节的十六进制来表示

**Reply by draveness** [reply]
我这里说的不是十六进制啊

**Reply by unbreakablehf** [reply]
你睡觉也太迟了， 注意身体！

**Comment by Dream4ever** [comment]
> @draveness
> > 感谢作者的文章，又长知识了。
> > 
> > PS：“Git 对每一个版本的提交都会生成一个 20 字节的 SHA-1 哈希”，这里应该是“40 字节”吧？
> 
> 是 20 字节啊 160bits == 20bytes https://en.wikipedia.org/wiki/SHA-1

哈哈，的确是我搞错了，把字节和字符串长度混为一谈了，又长知识了

**Comment by Cherrison** [comment]
> @xaclincoln
> 在工作中，没有直接感觉到Git的可靠性（SVN也挺可靠的），在分支操作上能够明显察觉Git的高性能。有一点不好的是，Git的工作流相对复杂，需要培训和一段时间的适应，很多新人难以掌握。

新人应该上岗之前就主动学习吧, 现在很多学生都能熟练使用git了

**Comment by draveness** [comment]
> > @xaclincoln
> > 在工作中，没有直接感觉到Git的可靠性（SVN也挺可靠的），在分支操作上能够明显察觉Git的高性能。有一点不好的是，Git的工作流相对复杂，需要培训和一段时间的适应，很多新人难以掌握。
> 
> 新人应该上岗之前就主动学习吧, 现在很多学生都能熟练使用git了

是啊，不过还是有些公司会用 SVN...

**Comment by songquanpeng** [comment]
> 由于 2017 年 GitHub 检测到了 SHA-1 的哈希碰撞

原文说的很清楚，并没有检测到，这里写错了。

> Are there Git collisions?
> Not yet. 

文章的意思是 Github 会检测 SHA-1 的哈希碰撞，而非检测到了。

**Comment by rainydew** [comment]
git感觉稍微不太好的地方就是工具链比较复杂，大部分是命令式的。其他地方调用其api的心智负担较重，各种特性耦合比较紧
一些测试平台的用例和低代码平台，一些文档例如google docs，版本控制(如有)都是自己实现的，且代价比较大。其实如果git的能力更open些，协议更通用的话，或许可以对接过来
